packages/yew/src/virtual_dom/vlist.rs

```diff
@@ -38,22 +47,30 @@ impl Deref for VList {
     type Target = Vec<VNode>;
 
     fn deref(&self) -> &Self::Target {
-        &self.children
+        match self.children {
+            Some(ref m) => m,
+            None => {
+                // This is mutable because the Vec<VNode> is not Sync
+                static mut EMPTY: Vec<VNode> = Vec::new();
+                // SAFETY: The EMPTY value is always read-only
+                unsafe { &EMPTY }
+            }
+        }
```

Comment:
This looks a bit tricky overall. Is it possible to achieve this while using safety?

I'm also not sure I understand why we need an Option around the children type. Probably it is related to this whole thing.

Can't we have a default static value like this?
```
static EMPTY: Rc<Vec<VNode>> = Rc::new(Vec::new())
```
(Maybe a thread-local one but you see the idea)

Reply:
Since VNode is `!Send`, this makes a global `static EMPTY: &Vec<VNode>= &Vec::new()` inaccessible from all threads. (Rust will not compile this anyways...)

If we use `thread_local!`, this variable will only live for the period of the thread, not the period of the program, hence not `'static`. This means that Rust cannot guarantee the reference will outlive the ownership. (Which is not quite right, since `!Send` types cannot live longer than the thread it resides on. So this is technically `'static` for them.) In this case, I think the only way is to teach the Rust compiler a lesson with some additional knowledge.

![image](https://user-images.githubusercontent.com/11693215/229351009-630ea114-3f1b-4db9-a693-2307eeaad64b.png)

However, we can still do it with safe Rust by leaking the memory.
If we don't have SSR, this might be the preferred method over unsafe since there is only 1 thread. But we also have SSR these days which will cause leaked memory for each thread ever created.

```rust
thread_local! {
    static EMPTY: &'static Vec<VNode> = Box::leak(Box::default());
}

EMPTY.with(|m| *m)
```

If there is a way to achieve this with safe Rust, I am also very curious to know.

Reply:
Maybe a stupid question but... do we actually need to impl Deref and DerefMut on VList??

The only use in its own crate is for this and it's kinda non-brainer to work around:

```patch
diff --git a/packages/yew/src/virtual_dom/vlist.rs b/packages/yew/src/virtual_dom/vlist.rs
index fc211ce9..40c94d5d 100644
--- a/packages/yew/src/virtual_dom/vlist.rs
+++ b/packages/yew/src/virtual_dom/vlist.rs
@@ -43,6 +43,7 @@ impl Default for VList {
     }
 }
 
+/*
 impl Deref for VList {
     type Target = Vec<VNode>;
 
@@ -58,13 +59,16 @@ impl Deref for VList {
         }
     }
 }
+*/
 
+/*
 impl DerefMut for VList {
     fn deref_mut(&mut self) -> &mut Self::Target {
         self.fully_keyed = FullyKeyedState::Unknown;
         self.children_mut()
     }
 }
+*/
 
 impl VList {
     /// Creates a new empty [VList] instance.
@@ -135,7 +139,7 @@ impl VList {
         match self.fully_keyed {
             FullyKeyedState::KnownFullyKeyed => true,
             FullyKeyedState::KnownMissingKeys => false,
-            FullyKeyedState::Unknown => self.iter().all(|c| c.has_key()),
+            FullyKeyedState::Unknown => self.children.iter().flat_map(|x| x.iter()).all(|c| c.has_key()),
         }
     }
 }
```

Reply:
I guess the original intention for implementing VList with dereferencing to `Vec<VNode>` is to allow users use all methods available to a vector to CURD VLists.

E.g.: In bounce, I used VList as a Vec to recursively read all available html to be used with the `<head />` element.
https://github.com/bounce-rs/bounce/blob/master/crates/bounce/src/helmet/comp.rs#L27
I didn't use any mutable operations in bounce helmet, but I can imagine a use case where manipulates VList would exist.